share/man/man4/isa.4
.\"	$NetBSD$
.\"
.\" Copyright (c) 1997 Jason R. Thorpe.  All rights reserved.
.\" Copyright (c) 1997 Jonathan Stone
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgements:
.\"      This product includes software developed by Jonathan Stone
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd June 10, 2013
.Dt ISA 4
.Os
.Sh NAME
.Nm isa
.Nd introduction to machine-independent ISA bus support and drivers
.Sh SYNOPSIS
Attachments are machine-dependent and depend on the bus topology and
.Tn ISA
bus interface of your system.
See
.Xr intro 4
for your system for details.
.Sh DESCRIPTION
.Nx
includes a machine-independent
.Tn ISA
bus subsystem and several machine-independent
.Tn ISA
device drivers.
.Pp
Your system may support additional
.Tn ISA
devices.
Drivers for
.Tn ISA
devices not listed here are machine-dependent.
Consult your system's
.Xr intro 4
for additional information.
.Sh SUPPORTED DEVICES
.Nx
includes machine-independent
.Tn ISA
drivers, sorted by device type and driver name:
.Ss SCSI interfaces
.Bl -tag -width pcdisplay -offset indent
.It adv
Advansys
.Tn SCSI
interfaces.
.It aha
Adaptec AHA-154x family (154xA, 154xB, 154xC, and 154xCF) and the
BusLogic BT54x
.Tn SCSI
interfaces.
.It ahc
Adaptec 29xx, 39xx, and other AIC-7xxx-based
.Tn SCSI
interfaces.
.It aic
Adaptec AIC-6260 and Adaptec AIC-6360 based
.Tn SCSI
interfaces, including
the Adaptec 152x, SoundBlaster
.Tn SCSI
interfaces, and a variety of compatibles.
.It bha
BusLogic BT-445
.Tn SCSI
interfaces.
.It dpt
DPT SmartCache/SmartRAID III and IV
.Tn SCSI
interfaces.
.It esp
NCR 53C9x, Emulex ESP406, and Qlogic FAS408
.Tn SCSI
interfaces.
.It nca
NCR-5380/NCR-53C400
.It sea
Seagate/Future Domain
.Tn SCSI
cards.
ST01/02, Future Domain TMC-885, and Future Domain TMC-950.
.It uha
Ultrastor 14f
.Tn SCSI
interfaces.
.It wds
WD-7000 family of bus-mastering
.Tn SCSI
interfaces.
.El
.Ss Disk and tape controllers
.Bl -tag -width pcdisplay -offset indent
.It mcd
Mitsumi CD-ROM drives.
.It wdc
Standard Western Digital type hard drive controllers: MFM, RLL, ESDI,
and IDE/ATAPI.
.It wt
Wangtek and compatible QIC-02 and QIC-36 tape drives.
.El
.Ss Serial and parallel interfaces
.Bl -tag -width pcdisplay -offset indent
.It ast
Multi-port serial communications card first made by AST.
.It boca
Boca BB100[48] and BB2016 multiplexing serial communications cards.
.It com
NS8250-, NS16450-, and NS16550-based serial ports.
.It cy
Cyclades Cyclom-4Y, -8Y, and -16Y asynchronous serial communications cards.
.It ioat
.Tn BOCA Research
IOAT66
serial interfaces.
.It lpt
Standard
.Tn ISA
parallel port interface.
.It rtfps
IBM RT four-port serial interfaces.
.It tcom
Byte Runner Technologies TC-400 and TC-800 series serial interfaces.
.El
.Ss Network interfaces
.Bl -tag -width pcdisplay -offset indent
.It ai
.Tn AT\*[Am]T
.Tn StarLan
.Tn Ethernet
interfaces.
.It ate
Allied Telesis AT1700 series and RE2000 series
.Tn Ethernet
interfaces.
.It cs
.Tn Cirrus Logic
Crystal CS8900
.Tn Ethernet
interfaces.
.It ec
3Com EtherLink II (3c503)
.Tn Ethernet
interfaces.
.It ef
.Tn 3Com
EtherLink II (3c507)
.Tn Ethernet
interfaces.
.It eg
3Com EtherLink Plus (3c505)
.Tn Ethernet
interfaces.
.It el
3Com EtherLink (3c501)
.Tn Ethernet
interfaces.
.It ep
3Com EtherLink III (3c509)
.Tn Ethernet
interfaces.
.It fmv
Fujitsu FMV-181 and FMV-182 interfaces.
.\".It hp
.\" Broken driver for HP -relabelled
.\" .Tn Ethernet
.\" cards.
.It ix
Intel EtherExpress/16
.Tn Ethernet
interfaces.
.It iy
Intel i82595-based
.Tn Ethernet
interfaces, including the EtherExpress Pro/10.
.It lc
DEC EtherWORKS III
.Tn Ethernet
interfaces (DE203, DE204, and DE205).
.It le
.Tn Ethernet
interfaces based on the AMD LANCE chip,
including BICC Isolan, Novell NE2100, Digital DEPCA, and PCnet-ISA.
.It ne
Novel NE2000 and compatible
.Tn Ethernet
interfaces.
.It ntwoc
SDL Communications Riscom/N2 synchronous serial interfaces.
.It sm
SMC91C9x-based
.Tn Ethernet
interfaces.
.It tr
TROPIC based token ring interfaces.
.It we
Western Digital/SMC 80x3, SMC Elite Ultra, and SMC EtherEZ
.Tn Ethernet
interfaces.
.El
.Ss Sound cards and MIDI interfaces
.Bl -tag -width pcdisplay -offset indent
.It aria
Sierra's Aria based sound cards.
.It cms
Creative Music System.
.It ess
ESS Technology AudioDrive 1788-, 1888-, 1887-, and 888-based sound cards.
.It gus
Gravis Ultrasound sound cards.
.It mpu
Roland MPU401 (and compatible) MIDI UARTs.
.It opl
Yamaha OPL2 and OPL3 FM MIDI synthesizers.
.It pas
ProAudio Spectrum sound cards.
.It sb
SoundBlaster, SoundBlaster 16, and SoundBlaster Pro sound cards.
.It wss
Windows Sound System-compatible sound cards based on the AD1848 and
compatible chips.
.El
.Ss Miscellaneous devices
.Bl -tag -width pcdisplay -offset indent
.It az
Aztech/PackardBell radio card.
.It ega
EGA graphics boards.
.It lm
National Semiconductor LM78, LM79 and compatible hardware monitors.
.It pcdisplay
PC display adapters.
.It pcic
.Tn PCI
.Tn PCMCIA
controllers, including the Cirrus Logic GD6729.
.It pckbc
PC keyboard controllers.
.It pcppi
PC control and timer ports.
.It pms
PS/2 auxiliary port mice (including wheel mice).
.It rt
AIMS Lab Radiotrack FM radio.
.It rtii
AIMS Lab Radiotrack II FM radio.
.It sf2r
SoundForte RadioLink SF16-FMR2 FM radio.
.It soekrisgpio
Soekris net6501 GPIO and LEDs driver
.It tcic
Databook DB86082, DB86084, DB86184, and DB86072
.Tn PCMCIA
controllers.
.It vga
VGA graphics boards.
.It wbsio
Winbond LPC Super I/O.
.El
.Pp
Note that some
.Tn ISA
devices also have newer
.Tn ISA
Plug-and-Play variants.
These are listed in
.Xr isapnp 4 .
Some i386 platforms use
.Xr pnpbios 4
to attach
.Tn ISA
devices.
.Sh DIAGNOSTICS
.Bl -diag
.It "Stray interrupt on IRQ 7"
It means the interrupt controller reported an unmasked interrupt on IRQ
7, but no driver attached to that IRQ `claimed' it.
.Pp
There are two reasons this can happen:
.Bl -bullet
.It
In anything other than i386, it would almost always mean that there is a
driver attached to the IRQ, but it is the wrong driver.
.It
On i386, there is the more obscure issue of `default IRQ7's.
That is, when a device asserts an IRQ, but the IRQ is deasserted
after the PIC latches the interrupt and before the CPU acknowledges
it, the PIC just flat out lies about which IRQ it was.
It is usually due to a suboptimally coded driver.
.El
.El
.Sh SEE ALSO
.Xr adv 4 ,
.Xr aha 4 ,
.Xr ahc 4 ,
.Xr ai 4 ,
.Xr aic 4 ,
.Xr aria 4 ,
.Xr ast 4 ,
.Xr ate 4 ,
.Xr az 4 ,
.Xr bha 4 ,
.Xr boca 4 ,
.Xr cms 4 ,
.Xr com 4 ,
.Xr cs 4 ,
.Xr cy 4 ,
.Xr dpt 4 ,
.Xr ec 4 ,
.Xr ef 4 ,
.Xr eg 4 ,
.Xr el 4 ,
.Xr ep 4 ,
.Xr esp 4 ,
.Xr ess 4 ,
.Xr fmv 4 ,
.Xr gus 4 ,
.Xr intro 4 ,
.Xr ioat 4 ,
.Xr isapnp 4 ,
.Xr ix 4 ,
.Xr iy 4 ,
.Xr joy 4 ,
.Xr lc 4 ,
.Xr le 4 ,
.Xr lm 4 ,
.Xr lpt 4 ,
.Xr mcd 4 ,
.Xr mpu 4 ,
.Xr nca 4 ,
.Xr ne 4 ,
.Xr ntwoc 4 ,
.Xr opl 4 ,
.Xr pas 4 ,
.Xr pcdisplay 4 ,
.Xr pcic 4 ,
.Xr pckbc 4 ,
.Xr pcppi 4 ,
.Xr pms 4 ,
.Xr pnpbios 4 ,
.Xr rt 4 ,
.Xr rtfps 4 ,
.Xr rtii 4 ,
.Xr sb 4 ,
.Xr sea 4 ,
.Xr sf2r 4 ,
.Xr sm 4 ,
.Xr soekrisgpio 4 ,
.Xr tcic 4 ,
.Xr tcom 4 ,
.Xr tr 4 ,
.Xr uha 4 ,
.Xr vga 4 ,
.Xr wbsio 4 ,
.Xr wd 4 ,
.Xr wdc 4 ,
.Xr wds 4 ,
.Xr we 4 ,
.Xr wss 4 ,
.Xr wt 4
.Sh HISTORY
The machine-independent
.Tn ISA
subsystem appeared in
.Nx 1.2 .


doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision$>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------


Changes from NetBSD 7.0 to NetBSD 8.0:
	libc: Update to tzcode2014f. [christos 20140815]
	gmake:	Update to gmake-3.81 in external/gpl2/gmake and remove 
		gnu/dist/gmake (3.80) [christos 20140818]
	module(7): Create MODULAR_DEFAULT_AUTOLOAD config flag to control
		whether module autoloading is on by default.
		[jnemeth 20140824]
	kernel: Option SYMTAB_SPACE has been replaced with COPY_SYMTAB.
		The related changes to src/sys/conf require a rebuild of
		config(1). [joerg 20140824]
	zoneinfo: Import tzdata2014g. [apb 20140831]
	wm(4): Add internal SERDES support for 82575 and newer controllers.
		[msaitoh 20140903]
	or1k: Add new port.  [matt 20140903]
	awindma(4): Add support for Allwinner A10/A20 DMA controller.
		[jmcneill 20140906]
	awinac(4): Add support for Allwinner A10/A20 audio codec.
		[jmcneill 20140906]
	awinrtc(4): Add support for Allwinner A10/A20 real-time clock.
		[jmcneill 20140907]
	awinmmc(4): Use DMA for MMC transfers. [jmcneill 20140908]
	axp20x(4): Add AXP20x PMU driver. [jmcneill 20140909]
	rtsol(8): Removed in favour of dhcpcd. [roy 20140911]
	rtsold(8): Removed in favour of dhcpcd. [roy 20140911]
	dhcpcd(8): Import dhcpcd-6.4.5. [roy 20140918]
	arm: Add support for i.MX6 SoC. [ryo 20140925]
	gpt(8): Completed overhaul, including adding follwing subcommands:
		resize, set, unset, backup, restore, and resizedisk.
		[jnemeth 20140926]
	dhcpcd(8): Import dhcpcd-6.4.7. [roy 20140927]
	mpt(4): added bio(4) support. [jmcneill 20140927]
	gpt(8): added type subcommand to modify partition type.
		[jnemeth 20140927]
	zoneinfo: Import tzdata2014h. [apb 20141004]
	dhcpcd(8): Import dhcpcd-6.5.0. [roy 20141006]
	libc: Update to tzcode2014h. [christos 20141007]
	file(1): Upgraded to 5.20. [christos 20141010]
	arm: Add support for Allwinner A31 SoC. [jmcneill 20141010]
	awiniic(4): Add support for Allwinner A31 SoC. [jmcneill 20141012]
	awinp2wi(4): Add support for Allwinner A31 Push-Pull Two Wire
		Interface. [jmcneill 20141012]
	axp22x(4): Add AXP22x PMU driver. [jmcneill 20141012]
	awinrtc(4): Add support for Allwinner A31 real-time clock.
		[jmcneill 20141012]
	awindma(4): Add support for Allwinner A31 DMA controller.
		[jmcneill 20141013]
	awinac(4): Add support for Allwinner A31 audio codec.
		[jmcneill 20141013]
	motg(4): Add support for Allwinner A10/A20 SoC. [jmcneill 20141015]
	wpa: Import wpa_supplicant and hostapd 2.3. [christos 20141016]
	dhcpcd(8): Import dhcpcd-6.5.1. [roy 20141018]
	OpenSSH: Imported 6.7. [christos 20141018]
	resolvconf(8): Import openresolv-3.6.0 [roy 20141020]
	gcc: Support -fsanitize=address [christos 20141022]
	zoneinfo: Import tzdata2014i. [apb 20141023]
	libc: Update to tzcode2014i. [christos 20141023]
	openpam(3): update to 20140912 (ourouparia) [christos 20141024]
	pppd(8): updated to version 2.4.7. [christos 20141025]
	acpi(4): Updated ACPICA to 20140926. [christos 20141025]
	dhcpcd(8): Import dhcpcd-6.6.0 [roy 20141029]
	resolvconf(8): Import openresolv-3.6.1 [roy 20141029]
	rtsx(4): Add support for Realtek RTS5227/RTL8402/RTL8411/RTL8411B.
		[nonaka 20141029]
	flex(1): Import flex-2.5.37 [christos 20141029]
	iwn(4): Add support for Intel Centrino Wireless-N
		100/105/130/135/2200/2230. From OpenBSD. [nonaka 20141030]
	awge(4): Add support for gigabit ethernet found on various
		Allwinner SOCs. [martin 20141101]
	dhcpcd(8): Import dhcpcd-6.6.1. [roy 20141107]
	awinhdmi(4): Add support for Allwinner A20/A31 HDMI controller.
		[jmcneill 20141110]
	awintcon(4): Add support for Allwinner A20/A31 TV/LCD controller.
		[jmcneill 20141110]
	awindebe(4): Add support for Allwinner A20/A31 Display engine
		backend (DE-BE). [jmcneill 20141110]
	awinhdmiaudio(4): Add support for Allwinner A20/A31 HDMI audio
		controller. [jmcneill 20141111]
	dhcpcd(8): Import dhcpcd-6.6.2. [roy 20141114]
	awinir(4): Add support for Allwinner A20/A31 IR receiver.
		[jmcneill 20141115]
	zoneinfo: Import tzdata2014j. [apb 20141116]
	libc: Update to tzcode2014j. [christos 20141117]
	libpcap: Import 1.6.2. [christos 20141119]
	tcpdump(8): Import 4.6.2. [christos 20141119]
	dhcpcd(8): Import dhcpcd-6.6.4. [roy 20141126]
	dhcpcd(8): Import dhcpcd-6.6.5. [roy 20141209]
	bind: Import version 9.10.1-P1. [christos 20141209]


sys/dev/isa/isa.c
/*	$NetBSD$	*/

/*-
 * Copyright (c) 1998, 2001, 2008 The NetBSD Foundation, Inc.
 * All rights reserved.
 *
 * This code is derived from software contributed to The NetBSD Foundation
 * by Charles M. Hannum; by Jason R. Thorpe of Wasabi Systems, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE NETBSD FOUNDATION, INC. AND CONTRIBUTORS
 * ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
 * TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR CONTRIBUTORS
 * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include <sys/cdefs.h>
__KERNEL_RCSID(0, "$NetBSD$");

#include <sys/param.h>
#include <sys/systm.h>
#include <sys/kernel.h>
#include <sys/malloc.h>
#include <sys/device.h>

#include <sys/intr.h>

#include <dev/isa/isareg.h>
#include <dev/isa/isavar.h>
#include <dev/isa/isadmareg.h>

#include "isadma.h"

#include "isapnp.h"
#if NISAPNP > 0
#include <dev/isapnp/isapnpreg.h>
#include <dev/isapnp/isapnpvar.h>
#endif

#include "locators.h"

int	isamatch(device_t, cfdata_t, void *);
void	isaattach(device_t, device_t, void *);
int	isadetach(device_t, int);
int	isarescan(device_t, const char *, const int *);
void	isachilddetached(device_t, device_t);
int	isaprint(void *, const char *);

CFATTACH_DECL2_NEW(isa, sizeof(struct isa_softc),
    isamatch, isaattach, isadetach, NULL, isarescan, isachilddetached);

void	isa_attach_knowndevs(struct isa_softc *);
void	isa_free_knowndevs(struct isa_softc *);

int	isasubmatch(device_t, cfdata_t, const int *, void *);
int	isasearch(device_t, cfdata_t, const int *, void *);

static int	isa_slotcount = -1;	/* -1 == don't know how many */

int
isamatch(device_t parent, cfdata_t cf, void *aux)
{
	/* XXX check other indicators */

	return (1);
}

void
isaattach(device_t parent, device_t self, void *aux)
{
	struct isa_softc *sc = device_private(self);
	struct isabus_attach_args *iba = aux;
	static const int wildcard[ISACF_NLOCS] = {
		ISACF_PORT_DEFAULT, ISACF_SIZE_DEFAULT,
		ISACF_IOMEM_DEFAULT, ISACF_IOSIZ_DEFAULT,
		ISACF_IRQ_DEFAULT, ISACF_DRQ_DEFAULT, ISACF_DRQ2_DEFAULT
	};

	TAILQ_INIT(&sc->sc_knowndevs);
	sc->sc_dynamicdevs = 0;

	sc->sc_dev = self;

	isa_attach_hook(parent, self, iba);
	aprint_naive("\n");
	aprint_normal("\n");

	/* XXX Add code to fetch known-devices. */

	sc->sc_iot = iba->iba_iot;
	sc->sc_memt = iba->iba_memt;
	sc->sc_dmat = iba->iba_dmat;
	sc->sc_ic = iba->iba_ic;

#if NISAPNP > 0
	/*
	 * Reset isapnp cards that the bios configured for us
	 */
	isapnp_isa_attach_hook(sc);
#endif

#if NISADMA > 0
	/*
	 * Initialize our DMA state.
	 */
	isa_dmainit(sc->sc_ic, sc->sc_iot, sc->sc_dmat, self);
#endif

	/* Attach all direct-config children. */
	isa_attach_knowndevs(sc);

	/*
	 * If we don't support dynamic hello/goodbye of devices,
	 * then free the knowndevs info now.
	 */
	if (sc->sc_dynamicdevs == 0)
		isa_free_knowndevs(sc);

	/* Attach all indirect-config children. */
	isarescan(self, "isa", wildcard);

	if (!pmf_device_register(self, NULL, NULL))
		aprint_error_dev(self, "couldn't establish power handler\n");
}

int
isadetach(device_t self, int flags)
{
	struct isa_softc *sc = device_private(self);
	int rc;

	if ((rc = config_detach_children(self, flags)) != 0)
		return rc;

	pmf_device_deregister(self);

	isa_free_knowndevs(sc);

#if NISADMA > 0
	isa_dmadestroy(sc->sc_ic);
#endif
	isa_detach_hook(sc->sc_ic, self);

	return 0;
}

int
isarescan(device_t self, const char *ifattr, const int *locators)
{
	prop_dictionary_t dict;
	int locs[ISACF_NLOCS];
	bool no_legacy_devices = false;

	dict = device_properties(self);
	if (prop_dictionary_get_bool(dict, "no-legacy-devices",
	    &no_legacy_devices) == true) {
		aprint_debug_dev(self, "platform reports no legacy devices\n");
		return 0;
	}

	memcpy(locs, locators, sizeof(locs));

	/*
	 * XXX Bus independent code calling this function does not
	 * know the locator default values. It assumes "-1" for now.
	 * (should be made available by "config" one day)
	 * So fixup where the "-1" is not correct.
	 */
	if (locs[ISACF_SIZE] == -1)
		locs[ISACF_SIZE] = ISACF_SIZE_DEFAULT;
	if (locs[ISACF_IOSIZ] == -1)
		locs[ISACF_IOSIZ] = ISACF_IOSIZ_DEFAULT;

	config_search_loc(isasearch, self, ifattr, locs, NULL);
	return (0);
}

void
isachilddetached(device_t self, device_t child)
{
	struct isa_knowndev *ik;
	struct isa_softc *sc = device_private(self);

	TAILQ_FOREACH(ik, &sc->sc_knowndevs, ik_list) {
		if (ik->ik_claimed == child)
			ik->ik_claimed = NULL;
	}
}

void
isa_attach_knowndevs(struct isa_softc *sc)
{
	struct isa_attach_args ia;
	struct isa_knowndev *ik;

	if (TAILQ_EMPTY(&sc->sc_knowndevs))
		return;

	TAILQ_FOREACH(ik, &sc->sc_knowndevs, ik_list) {
		if (ik->ik_claimed != NULL)
			continue;

		ia.ia_iot = sc->sc_iot;
		ia.ia_memt = sc->sc_memt;
		ia.ia_dmat = sc->sc_dmat;
		ia.ia_ic = sc->sc_ic;

		ia.ia_pnpname = ik->ik_pnpname;
		ia.ia_pnpcompatnames = ik->ik_pnpcompatnames;

		ia.ia_io = ik->ik_io;
		ia.ia_nio = ik->ik_nio;

		ia.ia_iomem = ik->ik_iomem;
		ia.ia_niomem = ik->ik_niomem;

		ia.ia_irq = ik->ik_irq;
		ia.ia_nirq = ik->ik_nirq;

		ia.ia_drq = ik->ik_drq;
		ia.ia_ndrq = ik->ik_ndrq;

		ia.ia_aux = NULL;

		/* XXX should setup locator array */

		ik->ik_claimed = config_found_sm_loc(sc->sc_dev,
		    "isa", 0, &ia, isaprint, isasubmatch);
	}
}

void
isa_free_knowndevs(struct isa_softc *sc)
{
	struct isa_knowndev *ik;
	struct isa_pnpname *ipn;

#define	FREEIT(x)	if (x != NULL) free(x, M_DEVBUF)

	while ((ik = TAILQ_FIRST(&sc->sc_knowndevs)) != NULL) {
		TAILQ_REMOVE(&sc->sc_knowndevs, ik, ik_list);
		FREEIT(ik->ik_pnpname);
		while ((ipn = ik->ik_pnpcompatnames) != NULL) {
			ik->ik_pnpcompatnames = ipn->ipn_next;
			free(ipn->ipn_name, M_DEVBUF);
			free(ipn, M_DEVBUF);
		}
		FREEIT(ik->ik_io);
		FREEIT(ik->ik_iomem);
		FREEIT(ik->ik_irq);
		FREEIT(ik->ik_drq);
		free(ik, M_DEVBUF);
	}

#undef FREEIT
}

static int
checkattachargs(struct isa_attach_args *ia, const int *loc)
{
	int i;

	if (ia->ia_nio == 0) {
		if (loc[ISACF_PORT] != ISACF_PORT_DEFAULT)
			return (0);
	} else {
		if (loc[ISACF_PORT] != ISACF_PORT_DEFAULT &&
		    loc[ISACF_PORT] != ia->ia_io[0].ir_addr)
			return (0);
	}

	if (ia->ia_niomem == 0) {
		if (loc[ISACF_IOMEM] != ISACF_IOMEM_DEFAULT)
			return (0);
	} else {
		if (loc[ISACF_IOMEM] != ISACF_IOMEM_DEFAULT &&
		    loc[ISACF_IOMEM] != ia->ia_iomem[0].ir_addr)
			return (0);
	}

	if (ia->ia_nirq == 0) {
		if (loc[ISACF_IRQ] != ISACF_IRQ_DEFAULT)
			return (0);
	} else {
		if (loc[ISACF_IRQ] != ISACF_IRQ_DEFAULT &&
		    loc[ISACF_IRQ] != ia->ia_irq[0].ir_irq)
			return (0);
	}

	if (ia->ia_ndrq == 0) {
		if (loc[ISACF_DRQ] != ISACF_DRQ_DEFAULT)
			return (0);
		if (loc[ISACF_DRQ2] != ISACF_DRQ2_DEFAULT)
			return (0);
	} else {
		for (i = 0; i < 2; i++) {
			if (i == ia->ia_ndrq)
				break;
			if (loc[ISACF_DRQ + i] != ISACF_DRQ_DEFAULT &&
			    loc[ISACF_DRQ + i] != ia->ia_drq[i].ir_drq)
				return (0);
		}
		for (; i < 2; i++) {
			if (loc[ISACF_DRQ + i] != ISACF_DRQ_DEFAULT)
				return (0);
		}
	}

	return (1);
}

int
isasubmatch(device_t parent, cfdata_t cf, const int *ldesc, void *aux)
{
	struct isa_attach_args *ia = aux;

	if (!checkattachargs(ia, cf->cf_loc))
		return (0);

	return (config_match(parent, cf, aux));
}

int
isaprint(void *aux, const char *isa)
{
	struct isa_attach_args *ia = aux;
	const char *sep;
	int i;

	/*
	 * This block of code only fires if we have a direct-config'd
	 * device for which there is no driver match.
	 */
	if (isa != NULL) {
		struct isa_pnpname *ipn;

		if (ia->ia_pnpname != NULL)
			aprint_normal("%s", ia->ia_pnpname);
		if ((ipn = ia->ia_pnpcompatnames) != NULL) {
			aprint_normal(" (");	/* ) */
			for (sep = ""; ipn != NULL;
			     ipn = ipn->ipn_next, sep = " ") {
				aprint_normal("%s%s", sep, ipn->ipn_name);
			}
	/* ( */		aprint_normal(")");
		}
		aprint_normal(" at %s", isa);
	}

	if (ia->ia_nio) {
		sep = "";
		aprint_normal(" port ");
		for (i = 0; i < ia->ia_nio; i++) {
			if (ia->ia_io[i].ir_size == 0)
				continue;
			aprint_normal("%s0x%x", sep, ia->ia_io[i].ir_addr);
			if (ia->ia_io[i].ir_size > 1)
				aprint_normal("-0x%x", ia->ia_io[i].ir_addr +
				    ia->ia_io[i].ir_size - 1);
			sep = ",";
		}
	}

	if (ia->ia_niomem) {
		sep = "";
		aprint_normal(" iomem ");
		for (i = 0; i < ia->ia_niomem; i++) {
			if (ia->ia_iomem[i].ir_size == 0)
				continue;
			aprint_normal("%s0x%x", sep, ia->ia_iomem[i].ir_addr);
			if (ia->ia_iomem[i].ir_size > 1)
				aprint_normal("-0x%x", ia->ia_iomem[i].ir_addr +
				    ia->ia_iomem[i].ir_size - 1);
			sep = ",";
		}
	}

	if (ia->ia_nirq) {
		sep = "";
		aprint_normal(" irq ");
		for (i = 0; i < ia->ia_nirq; i++) {
			if (ia->ia_irq[i].ir_irq == ISACF_IRQ_DEFAULT)
				continue;
			aprint_normal("%s%d", sep, ia->ia_irq[i].ir_irq);
			sep = ",";
		}
	}

	if (ia->ia_ndrq) {
		sep = "";
		aprint_normal(" drq ");
		for (i = 0; i < ia->ia_ndrq; i++) {
			if (ia->ia_drq[i].ir_drq == ISACF_DRQ_DEFAULT)
				continue;
			aprint_normal("%s%d", sep, ia->ia_drq[i].ir_drq);
			sep = ",";
		}
	}

	return (UNCONF);
}

int
isasearch(device_t parent, cfdata_t cf, const int *slocs, void *aux)
{
	struct isa_io res_io[1];
	struct isa_iomem res_mem[1];
	struct isa_irq res_irq[1];
	struct isa_drq res_drq[2];
	struct isa_softc *sc = device_private(parent);
	struct isa_attach_args ia;
	int flocs[ISACF_NLOCS];
	int tryagain;

	do {
		ia.ia_pnpname = NULL;
		ia.ia_pnpcompatnames = NULL;

		res_io[0].ir_addr = cf->cf_loc[ISACF_PORT];
		res_io[0].ir_size = 0;

		res_mem[0].ir_addr = cf->cf_loc[ISACF_IOMEM];
		res_mem[0].ir_size = cf->cf_loc[ISACF_IOSIZ];

		res_irq[0].ir_irq =
		    cf->cf_loc[ISACF_IRQ] == 2 ? 9 : cf->cf_loc[ISACF_IRQ];

		res_drq[0].ir_drq = cf->cf_loc[ISACF_DRQ];
		res_drq[1].ir_drq = cf->cf_loc[ISACF_DRQ2];

		ia.ia_iot = sc->sc_iot;
		ia.ia_memt = sc->sc_memt;
		ia.ia_dmat = sc->sc_dmat;
		ia.ia_ic = sc->sc_ic;

		ia.ia_io = res_io;
		ia.ia_nio = 1;

		ia.ia_iomem = res_mem;
		ia.ia_niomem = 1;

		ia.ia_irq = res_irq;
		ia.ia_nirq = 1;

		ia.ia_drq = res_drq;
		ia.ia_ndrq = 2;

		if (!checkattachargs(&ia, slocs))
			return (0);

		tryagain = 0;
		if (config_match(parent, cf, &ia) > 0) {
			/*
			 * This is not necessary for detach, but might
			 * still be useful to collect device information.
			 */
			flocs[ISACF_PORT] = ia.ia_io[0].ir_addr;
			flocs[ISACF_SIZE] = ia.ia_io[0].ir_size;
			flocs[ISACF_IOMEM] = ia.ia_iomem[0].ir_addr;
			flocs[ISACF_IOSIZ] = ia.ia_iomem[0].ir_size;
			flocs[ISACF_IRQ] = ia.ia_irq[0].ir_irq;
			flocs[ISACF_DRQ] = ia.ia_drq[0].ir_drq;
			flocs[ISACF_DRQ2] = ia.ia_drq[1].ir_drq;
			config_attach_loc(parent, cf, flocs, &ia, isaprint);
			tryagain = (cf->cf_fstate == FSTATE_STAR);
		}
	} while (tryagain);

	return (0);
}

const char *
isa_intr_typename(int type)
{

	switch (type) {
	case IST_NONE:
		return ("none");
	case IST_PULSE:
		return ("pulsed");
	case IST_EDGE:
		return ("edge-triggered");
	case IST_LEVEL:
		return ("level-triggered");
	default:
		panic("isa_intr_typename: invalid type %d", type);
	}
}

int
isa_get_slotcount(void)
{

	return isa_slotcount;
}

void
isa_set_slotcount(int arg)
{

	isa_slotcount = arg;
}
